'flutter apk' now supports dynamically registered services.

mpcomplete opened the pull request:
Third-party libraries can now provide their own mojo services. They do
so by adding a config.yaml file to their pub package which contains
- a list of service names and java classes which handles that service's
  registration.
- a list of pre-built .jar files to statically link with the app's shell
  when building the app.

mpcomplete commented on packages/gcm/lib/apk/config.yaml:
```diff
@@ -0,0 +1,11 @@
+  - jars/android_support_v7_mediarouter_java.dex.jar
+  - jars/gcm_lib.dex.jar
+  - jars/google_play_services_default_java.dex.jar
+  - jars/interfaces_java.dex.jar
```
It's probably better to put these jars into google cloud store or something, rather than bundling into the pub package. Unless you have a better suggestion.

abarth replied:
Where did these jars come from?  Do we have the right to redistribute them?

Hixie commented on examples/fitness/apk/AndroidManifest.xml:
```diff
@@ -18,7 +18,7 @@
     <!-- end for GCM -->
 
-    <application android:icon="@mipmap/ic_launcher" android:label="Fitness" android:name="org.domokit.fitness.FitnessApplication">
+    <application android:icon="@mipmap/ic_launcher" android:label="Fitness" android:name="org.domokit.sky.shell.SkyApplication">
```
what are the implications of this change?

mpcomplete replied:
The only thing FitnessApplication did was register the GCM mojo service. This is now done by the gcm service itself via this new dynamic registration system.

Hixie commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -151,14 +180,41 @@ class ApkCommand extends FlutterCommand {
     List<String> artifactPaths;
     if (runner.enginePath != null) {
       androidSdkPath = '${runner.enginePath}/third_party/android_tools/sdk';
+      // TODO(mpcomplete): This might be messier than necessary. Questions:
```
please resolve the questions before checking this in.

Hixie commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -170,6 +226,7 @@ class ApkCommand extends FlutterCommand {
       List<ArtifactType> artifactTypes = <ArtifactType>[
         ArtifactType.androidIcuData,
         ArtifactType.androidClassesDex,
+        'TODO(mpcomplete): we need to upload new artifacts',
```
?

mpcomplete replied:
I need to fix this before checking in. These artifacts are downloaded from cloud storage. The ones currently in storage won't work with this new method, so I need new ones.

Hixie commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -183,11 +240,13 @@ class ApkCommand extends FlutterCommand {
-    components.classesDex = new File(artifactPaths[1]);
-    components.libSkyShell = new File(artifactPaths[2]);
-    components.debugKeystore = new File(artifactPaths[3]);
+    components.jars = [new File(artifactPaths[1]), new File(artifactPaths[2])];
```
type the literal

Hixie commented:
What are the licenses on those binary files? We shouldn't be checking anything in that isn't completely open source.

Hixie commented:
cc @abarth since this seems like something he'd have opinions on.

abarth commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -64,12 +69,27 @@ class _ApkBuilder {
+    // copying chromium's build system.
+    List<String> packageArgs = [_dx.path,
+      '--dex',
+      '--force-jumbo',
```
We should figure out what this arg does...  It sounds scary.

mpcomplete replied:
I found http://tools.android.com/download/adt-21-preview which suggests it just allows more strings in the binary.

abarth commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -151,14 +180,41 @@ class ApkCommand extends FlutterCommand {
+      return;
+
+    for (String service in manifest['services']) {
+      File serviceRoot = '${ArtifactStore.packageRoot}/$service/apk';
```
This looks like a type error.  The right-hand side is a string and the left-hand side claims it's a File.

mpcomplete replied:
Yep, thanks.

abarth commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -151,14 +180,41 @@ class ApkCommand extends FlutterCommand {
+      // link classes.dex.jar + services.jar instead. (Maybe classes.dex is
+      // like an exe and classes.dex.jar is like a static lib?)
+      // 2. Is there a way to get sky/shell/BUILD.gn to generate
+      // classes.dex.jar as one file, instead of shell_lib.dex.jar + shell.jar?
```
Probably we should have the engine repro produce a single jar and run dex on it here with all the other jars.

mpcomplete replied:
Done.

abarth commented on packages/gcm/lib/apk/config.yaml:
```diff
@@ -0,0 +1,11 @@
+registration-class: org.domokit.gcm.RegistrationIntentService$MojoService
+jars:
```
These blobs need to go somewhere else.  GCM is not open source and cannot land in this open-source repository.

mpcomplete replied:
4 of the jars are from the android SDK. I realized I can just use the ones in developer's SDK (which can be referenced by the $androidSdk var). The other 2 (gcm_lib and interfaces_java) are built by flutter engine.

abarth commented:
IMHO, we should break this patch up into pieces.  For example, we can land the changes to the "apk" command and then migrate some trivial service (e.g., sensors) to using the new mechanism.  Then we can tech up to GCM.

mpcomplete commented on packages/gcm/lib/apk/config.yaml:
```diff
@@ -0,0 +1,13 @@
+services:
+  - name: gcm::GcmService
```
This is the name of the service as given in the gcm.mojom file (see https://github.com/flutter/engine/blob/master/sky/services/gcm/gcm.mojom ). This is a bit redundant. If we could annotate the mojom file directly, we could probably avoid this section.

mpcomplete commented:
This is ready for another look. One remaining task is to update the class.dex file in the artifact store. I need to find out how to do that.

abarth commented:
This looks good except I don't think we should check the jars into this repo.  Instead, we should download them from Google Storage, similar to how we download the other binary artifacts produced by the engine repository.

mpcomplete commented:
OK, ready for another look. I've updated this to download the jars from cloud storage. The gcm URLs are not live yet (I tested it with a local server), so I'll have to wait until the buildbot finishes uploading the artifacts,

abarth commented on packages/flutter_tools/lib/src/commands/apk.dart:
```diff
@@ -151,6 +179,33 @@ class ApkCommand extends FlutterCommand {
+      for (String jar in serviceConfig['jars']) {
+        // Jar might refer to an android SDK jar, or URL to download.
+        jar = jar.replaceAll('\$androidSdk', '${components.androidSdk.path}');
+        if (jar.startsWith("http")) {
```
I wonder if we want to change `$androidSdk` into something like `android-sdk:` so that all the entries in this list are URLs of some form.

abarth commented on packages/gcm/lib/apk/config.yaml:
```diff
@@ -0,0 +1,13 @@
+  - $androidSdk/extras/google/google_play_services/libproject/google-play-services_lib/libs/google-play-services.jar
+  - $androidSdk/extras/android/support/v13/android-support-v13.jar
+  - $androidSdk/extras/android/support/v7/appcompat/libs/android-support-v7-appcompat.jar
+  - $androidSdk/extras/android/support/v7/mediarouter/libs/android-support-v7-mediarouter.jar
```
E.g., this would be:

`- android-sdk:extras/android/support/v7/mediarouter/libs/android-support-v7-mediarouter.jar`

mpcomplete replied:
Good idea. Done.

abarth commented:
LGTM

mpcomplete committed:
'flutter apk' now supports dynamically registered services.

Third-party libraries can now provide their own mojo services. They do
so by adding a config.yaml file to their pub package which contains
- a list of service names and java classes which handles that service's
  registration.
- a list of pre-built .jar files to statically link with the app's shell
  when building the app.

eseidelGoogle commented:
Looks like this landed.  Can I remove these:
https://github.com/flutter/buildbot/blob/master/travis/build.sh#L20

mpcomplete commented:
Yes - I don't think I needed them anyway, since we fetch from a fixed revision.
